include/fiction/technology/sidb_dynamic_gate_library.hpp

```diff
@@ -0,0 +1,1244 @@
+     *
+     * @tparam CellLyt Cell-level layout type.
+     * @param lyt Cell-level layout
+     * @return A 2D array of characters representing the cell-level layout.
+     */
+    template <typename Lyt>
+    [[nodiscard]] static std::array<std::array<char, gate_x_size()>, gate_y_size()>
+    cell_level_layout_to_list(const Lyt& lyt)
+    {
+        std::array<std::array<char, gate_x_size()>, gate_y_size()> result{};
+        const auto                                                 all_cell =
+            all_sidbs_in_spanned_area({0, 0, 0}, to_siqad_coord(cell<Lyt>{gate_x_size() - 1, gate_y_size() - 1}));
+        uint64_t counter = 0;
+
+        for (auto& row : result)
+        {
+            for (auto& cell : row)
+            {
+                const auto cell_type = lyt.get_cell_type(all_cell[counter]);
+                cell                 = (cell_type == Lyt::technology::cell_type::NORMAL ||
+                        cell_type == Lyt::technology::cell_type::OUTPUT) ?
+                                           (all_cell[counter].y != 20 || all_cell[counter].z != 0) ? 'x' : ' ' :
+                                           ' ';
+                counter++;
+            }
+        }
+        return result;
+    }
+    /**
+     * Designs an SiDB gate implementation if possible.
+     *
+     * @tparam GateLyt Pointy-top hexagonal gate-level layout type.
+     * @tparam CellLyt The type of the cell-level layout.
+     * @tparam GateLibraryblack Type of the gate-level layout used to generate the blacklist.
+     * @param skeleton Skeleton with atomic defects if available.
+     * @param t A vector of truth tables representing the logic functions for the gate.
+     * @param params Parameters for the SiDB gate design process.
+     * @param black_list A blacklist to avoid gate placements on certain ports.
+     * @param p The list of ports and their directions.
+     * @param tile The specific tile on which the gate should be designed.
+     *
+     * @return A fcn gate object, or an error fcn gate if no suitable layout is found.
+     */
+    template <typename Lyt, typename GateLyt, typename GateLibrary>
+    [[nodiscard]] static fcn_gate design_gate(
+        const Lyt& skeleton, const std::vector<tt>& t, const design_sidb_gates_params& params,
+        surface_black_list<
+            GateLyt, typename decltype(GateLibrary::get_gate_ports())::mapped_type::value_type::port_type>& black_list,
+        const port_list<port_direction>& p, const tile<GateLyt>& tile)
+    {
+        if (t == create_crossing_wire_tt() || t == create_double_wire_tt())
+        {
+            if (is_gate_design_impossible(
+                    skeleton, t, is_operational_params{params.phys_params, sidb_simulation_engine::QUICKEXACT}))
+            {
+                black_list[tile][create_id_tt()].push_back(p);
+                throw std::logic_error("Gate cannot be designed.");
```

Comment:
Instead of using pre-defined exceptions, rather make your own. It's more specific and helps you cover exactly what you need. See `fcn_gate_library` for examples.

Then, switch from `#include <stdexcept>` to `#include <exception>`.